### ISSUE TITLE
stress predict

### ISSUE THREAD

**Issue by NishmithaBashaboina** [NishmithaBashaboina/PREDICTIVE-STRESS-ANALYSIS#1]
import os
import zipfile

# Extract the zip archive
general_path = '/content/archive (5).zip'
with zipfile.ZipFile(general_path, 'r') as zip_ref:
    zip_ref.extractall("D:\facedata")  # Extract to a new folder

general_path = "D:\facedata" #Corrected path syntax
path_train_images = os.path.join(general_path, "train")  # Path to training images
path_test_images = os.path.join(general_path, "test")  # Path to test images

import tensorflow as tf

import keras
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, AveragePooling2D
from keras.layers import Dense, Activation, Dropout, Flatten

from keras.preprocessing import image
from tensorflow.keras.preprocessing.image import ImageDataGenerator
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

import tensorflow as tf     #primary deep learning library
from tensorflow.keras.preprocessing.image import ImageDataGenerator #for augmentation and image loading

#below are not needed if using transfer learning, needed to build network from scratch
# from tensorflow.keras.layers import Dense,Dropout,Conv2D,Flatten,MaxPooling2D,BatchNormalization
from tensorflow.keras.layers import Dense,Dropout,BatchNormalization

import os             # used for working with paths of images
import cv2

general_path = "D:\facedata" #Corrected path syntax
path_train_images = os.path.join(general_path, "train")  # Path to training images
path_test_images = os.path.join(general_path, "test")  # Path to test images
print("Training categories", os.listdir(path_train_images))
print("Testing categories", os.listdir(path_test_images))

Training categories ['sad', 'disgusted', 'neutral', 'happy', 'angry']
Testing categories ['sad', 'disgusted', 'neutral', 'happy', 'angry']

import matplotlib.pyplot as plt

train_base_dir=path_train_images
fig=plt.figure(figsize=(8,8))
i=1
for category in category_list:                                                          #pick one category out of all
    print("Visualizing category : ",category,"--------------------------------------")
    category_path=os.path.join(train_base_dir,category)     #using os to create the path
    flag=False                                              #using false, later changing value to break out of loop
    for img_name in os.listdir(category_path):              #pick img_name from what are available in category folder
        if(flag):
            break                                           #checking if this is the first time
#         print(os.path.join(category_path,img_name))
        img_path=os.path.join(category_path,img_name)       #getting img_path by adding img_name and category_path
        img_bgr=cv2.imread(img_path)                        #reading img by opencv, default read in BGR format
        print(img_path)
        img_rgb=cv2.cvtColor(img_bgr,cv2.COLOR_BGR2RGB)
        plt.subplot(1,7,i)                                  #subplot in a row(we already know it has 7 categories)
        plt.imshow(img_rgb)                                 #show img
        plt.title(category)                                 #add title
        flag=True                                           #changes val of flag so next time it breaks the loop
        i=i+1
plt.show() # display the result
Visualizing category :  sad --------------------------------------
D:acedata/train/sad/im3240.png
Visualizing category :  disgusted --------------------------------------
D:acedata/train/disgusted/im389.png
Visualizing category :  neutral --------------------------------------
D:acedata/train/neutral/im3240.png
Visualizing category :  happy --------------------------------------
D:acedata/train/happy/im6441.png
Visualizing category :  angry --------------------------------------
D:acedata/train/angry/im3240.png
#creating the imagedatagenerator, used to read and augment images
train_data_gen=ImageDataGenerator(
    rotation_range=10,
    width_shift_range=0.2,
    height_shift_range=0.2,
    shear_range=0.2,
    zoom_range=0.2,
    fill_mode='nearest',
    horizontal_flip=True,
    rescale=1./255)
train_data=train_data_gen.flow_from_directory(
    directory=train_base_dir,
    target_size=(224, 224),           #taking size as (224,224) as while using our fine tuning the arch has input shape as (224,244)
    class_mode='sparse',
    batch_size=64,
#     color_mode="grayscale"         # NO grayscale, the arch need 3 channel images ->(224,224,3)
)

Found 21087 images belonging to 5 classes.

train_data

<keras.src.legacy.preprocessing.image.DirectoryIterator at 0x7e3c69e7a920>
test_data=test_data_gen.flow_from_directory(
    directory=path_test_images,
    target_size=(224, 224),          #taking size as (224,224) as while using our fine tuning the arch has input shape as (224,244)
    class_mode='sparse',
    batch_size=64,                   # NO grayscale, the arch need 3 channel images ->(224,224,3)
#     color_mode="grayscale"
)
Found 3838 images belonging to 5 classes.

class_indices = train_data.class_indices
category_mapping = {v: k for k, v in class_indices.items()}
category_mapping
{0: 'angry', 1: 'disgusted', 2: 'happy', 3: 'neutral', 4: 'sad'}
model=tf.keras.applications.MobileNetV3Large()
model.summary()

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)              ┃ Output Shape           ┃        Param # ┃ Connected to           ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━┩
│ input_layer_1             │ (None, None, None, 3)  │              0 │ -                      │
│ (InputLayer)              │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ rescaling_1 (Rescaling)   │ (None, None, None, 3)  │              0 │ input_layer_1[0][0]    │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ conv (Conv2D)             │ (None, None, None, 16) │            432 │ rescaling_1[0][0]      │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ conv_bn                   │ (None, None, None, 16) │             64 │ conv[0][0]             │
│ (BatchNormalization)      │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ activation_21             │ (None, None, None, 16) │              0 │ conv_bn[0][0]          │
│ (Activation)              │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ expanded_conv_depthwise   │ (None, None, None, 16) │            144 │ activation_21[0][0]    │
│ (DepthwiseConv2D)         │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ expanded_conv_depthwise_… │ (None, None, None, 16) │             64 │ expanded_conv_depthwi… │
│ (BatchNormalization)      │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ re_lu_19 (ReLU)           │ (None, None, None, 16) │              0 │ expanded_conv_depthwi… │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ expanded_conv_project     │ (None, None, None, 16) │            256 │ re_lu_19[0][0]         │
│ (Conv2D)                  │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ expanded_conv_project_bn  │ (None, None, None, 16) │             64 │ expanded_conv_project… │
│ (BatchNormalization)      │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ expanded_conv_add (Add)   │ (None, None, None, 16) │              0 │ activation_21[0][0],   │
│                           │                        │                │ expanded_conv_project… │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ expanded_conv_1_expand    │ (None, None, None, 64) │          1,024 │ expanded_conv_add[0][… │
│ (Conv2D)                  │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ expanded_conv_1_expand_bn │ (None, None, None, 64) │            256 │ expanded_conv_1_expan… │
│ (BatchNormalization)      │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ re_lu_20 (ReLU)           │ (None, None, None, 64) │              0 │ expanded_conv_1_expan… │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ expanded_conv_1_depthwis… │ (None, None, None, 64) │              0 │ re_lu_20[0][0]         │
│ (ZeroPadding2D)           │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ expanded_conv_1_depthwise │ (None, None, None, 64) │            576 │ expanded_conv_1_depth… │
│ (DepthwiseConv2D)         │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ expanded_conv_1_depthwis… │ (None, None, None, 64) │            256 │ expanded_conv_1_depth… │
│ (BatchNormalization)      │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ re_lu_21 (ReLU)           │ (None, None, None, 64) │              0 │ expanded_conv_1_depth… │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ expanded_conv_1_project   │ (None, None, None, 24) │          1,536 │ re_lu_21[0][0]         │
│ (Conv2D)                  │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ expanded_conv_1_project_… │ (None, None, None, 24) │             96 │ expanded_conv_1_proje… │
│ (BatchNormalization)      │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ expanded_conv_2_expand    │ (None, None, None, 72) │          1,728 │ expanded_conv_1_proje… │
│ (Conv2D)                  │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ expanded_conv_2_expand_bn │ (None, None, None, 72) │            288 │ expanded_conv_2_expan… │
│ (BatchNormalization)      │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ re_lu_22 (ReLU)           │ (None, None, None, 72) │              0 │ expanded_conv_2_expan… │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ expanded_conv_2_depthwise │ (None, None, None, 72) │            648 │ re_lu_22[0][0]         │
│ (DepthwiseConv2D)         │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ expanded_conv_2_depthwis… │ (None, None, None, 72) │            288 │ expanded_conv_2_depth… │
│ (BatchNormalization)      │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ re_lu_23 (ReLU)           │ (None, None, None, 72) │              0 │ expanded_conv_2_depth… │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ expanded_conv_2_project   │ (None, None, None, 24) │          1,728 │ re_lu_23[0][0]         │
│ (Conv2D)                  │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ expanded_conv_2_project_… │ (None, None, None, 24) │             96 │ expanded_conv_2_proje… │
│ (BatchNormalization)      │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ expanded_conv_2_add (Add) │ (None, None, None, 24) │              0 │ expanded_conv_1_proje… │
│                           │                        │                │ expanded_conv_2_proje… │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ expanded_conv_3_expand    │ (None, None, None, 72) │          1,728 │ expanded_conv_2_add[0… │
│ (Conv2D)                  │                        │                │                        │
                

Total params: 5,507,432 (21.01 MB)
 Trainable params: 5,483,032 (20.92 MB)
 Non-trainable params: 24,400 (95.31 KB)
for layer in model.layers[:20]:
    layer.trainable = False
#for visualization ploting the model
tf.keras.utils.plot_model(model)
dot: graph is too large for cairo-renderer bitmaps. Scaling by 0.694628 to fit
#total number of layers in the model
len(model.layers)
194
#layer 1->index 0
model.layers[0]
<InputLayer name=input_layer_1, built=True>
model.layers[-1].output
<KerasTensor shape=(None, 1000), dtype=float32, sparse=False, name=keras_tensor_423>
#adding some ANN layers at last
output=model.layers[-1].output
output=BatchNormalization()(output)
output=Dense(128)(output)
output=BatchNormalization()(output)
# output=Dense(64,activation="relu")(output)
output=Dense(7,activation="softmax")(output)
#saving new_model
new_model=tf.keras.Model(inputs=model.input, outputs=output)
# fine tuning needs smaller learning rates !!
from keras.optimizers import Adam
adam=Adam(learning_rate=0.0001)
#compiling the model
new_model.compile(
    optimizer=adam,
    loss="sparse_categorical_crossentropy",
    metrics=["accuracy"])
from tensorflow.keras.callbacks import EarlyStopping,ReduceLROnPlateau
#used to stop early if no improvement on test/validation data-> prevent overfitting
earlystop=EarlyStopping(
    monitor='val_accuracy',
    patience=10,
    mode='min',
    restore_best_weights=True,
)
#reducing lear
reduce_lr = ReduceLROnPlateau(
    monitor='val_accuracy',
    factor=0.2,
    patience=3,
)
import warnings
warnings.filterwarnings('ignore')
history=new_model.fit(
    train_data,
    epochs=50,
    callbacks=[earlystop,reduce_lr],
    validation_data=test_data
    #steps_per_epochs=150
)

import pickle

# Example dictionary
my_dict = {0: 'happy', 1: 'angry', 2: 'sad', 3: 'neutral', 4: 'disgusted'}

# Define the file path
file_path = 'C:\\Users\\Shreya\\Downloads\\facedata.pkl'

# Write the dictionary to a file using pickle
with open(file_path, 'wb') as f:
    pickle.dump(category_mapping, f)

print("Dictionary exported to", file_path)
import pickle

# Example dictionary
my_dict = {0: 'happy', 1: 'angry', 2: 'sad', 3: 'neutral', 4: 'disgusted'}

# Define the file path
file_path = 'C:\\Users\\Shreya\\Downloads\\facedata.pkl'

# Write the dictionary to a file using pickle
with open(file_path, 'wb') as f:
    pickle.dump(category_mapping, f)

print("Dictionary exported to", file_path)
filname = '../content/archive (5).zip'
label_map = ['happy', 'angry', 'sad','neutral', 'disgusted']
def getData(filname):
    # images are 48x48
    # N = 35887
    Y = []
    X = []
    first = True
    for line in open(filname):
        if first:
            first = False
        else:
            row = line.split(',')
            Y.append(int(row[0]))
            X.append([int(p) for p in row[1].split()])

    X, Y = np.array(X) / 255.0, np.array(Y)
    return X, Y
objects = ('happy', 'angry', 'sad','neutral', 'disgusted')
y_pos = np.arange(len(objects))
print(y_pos)
[0 1 2 3 4]
def emotion_analysis(emotions):
    objects = ['angry', 'disgusted', 'happy', 'sad', 'neutral']
    aggregated_emotions = emotions[:5]
    y_pos = np.arange(len(objects))
    #emotions=np.mean(emotions, axis=1)
    plt.bar(y_pos, aggregated_emotions, align='center', alpha=0.9)
    plt.tick_params(axis='x', which='both', pad=10,width=4,length=10)
    plt.xticks(y_pos, objects)
    plt.ylabel('percentage')
    plt.title('emotion')

plt.show()
# Import necessary libraries
from IPython.display import display, Javascript
from google.colab.output import eval_js
from base64 import b64decode
import cv2
import numpy as np
from PIL import Image
import io
import matplotlib.pyplot as plt

# Import necessary libraries
from IPython.display import display, Javascript
from google.colab.output import eval_js
from base64 import b64decode
import cv2
import numpy as np
from PIL import Image
import io
import matplotlib.pyplot as plt

# JavaScript to capture an image from the webcam
def take_photo(filename='photo.jpg', quality=0.8):
    js = Javascript('''
    async function takePhoto(quality) {
        const div = document.createElement('div');
        const capture = document.createElement('button');
        capture.textContent = 'Capture';
        div.appendChild(capture);

        const video = document.createElement('video');
        video.style.display = 'block';
        const stream = await navigator.mediaDevices.getUserMedia({video: true});

        document.body.appendChild(div);
        div.appendChild(video);
        video.srcObject = stream;
        await video.play();

        // Resize the output video to a smaller size
        google.colab.output.setIframeHeight(document.documentElement.scrollHeight, true);

        // Wait for the capture button to be clicked.
        await new Promise((resolve) => capture.onclick = resolve);

        const canvas = document.createElement('canvas');
        canvas.width = video.videoWidth;
        canvas.height = video.videoHeight;
        canvas.getContext('2d').drawImage(video, 0, 0);
        stream.getTracks().forEach(track => track.stop());
        div.remove();

        return canvas.toDataURL('image/jpeg', quality);
    }
    ''')
    display(js)
    data = eval_js('takePhoto({})'.format(quality))
    binary = b64decode(data.split(',')[1])
    with open(filename, 'wb') as f:
        f.write(binary)
    return filename

# Capture and display the photo
filename = take_photo()
display(Image.open(filename))

from skimage import io
from tensorflow.keras.preprocessing import image
img = image.load_img('../content/photo.jpg', color_mode='rgb', target_size=(224, 224))
show_img=image.load_img('../content/photo.jpg', color_mode='rgb', target_size=(200, 200))
x = image.img_to_array(img)
x = np.expand_dims(x, axis = 0)

x /= 255

custom = model.predict(x)
#print(custom[0])
emotion_analysis(custom[0])

x = np.array(x, 'float32')
#x = x.reshape([48, 48]);

plt.gray()
plt.imshow(show_img)
plt.show()

m=0.000000000000000000001
a=custom[0]
ind=np.argmax(a[:len(objects)])
for i in range(0,len(a)):
    if a[i]>m:
        m=a[i]
        ind=i
ind = ind % len(objects)
print('Expression Prediction:',objects[ind])

Expression Prediction: disgusted
